osu.Game/Beatmaps/BeatmapDifficultyCache.cs

```diff
@@ -310,7 +310,7 @@ public DifficultyCacheLookup([NotNull] BeatmapInfo beatmap, [CanBeNull] RulesetI
                 Beatmap = beatmap;
                 // In the case that the user hasn't given us a ruleset, use the beatmap's default ruleset.
                 Ruleset = ruleset ?? Beatmap.Ruleset;
-                OrderedMods = mods?.OrderBy(m => m.Acronym).ToArray() ?? Array.Empty<Mod>();
+                OrderedMods = mods?.OrderBy(m => m.Acronym).Select(mod => mod.DeepClone()).ToArray() ?? Array.Empty<Mod>();
```

Comment:
Note that this would actually cause a deep clone of each mod in each single lookup, which means that every single setting value change and every combination of the mods applied would keep cloning mods to their respective lookups. This *may* become bad over time, although I haven't, nor can I profile this currently, so just keeping this as a note.

Reply:
Correct, which is why I was doubting the idea of caching setting-dependent difficulty values in the first place. But now that it's here something like this has to happen for it to work correctly. As stated in the OP I can copy the values out to a more light-weight structure but the fact remains that the data *has* to be there and be immutable for the cache to work.

Reply:
Ah so that's what your intention were when you attempted to switch to using an `APIMod` instead. I think it would be best to switch to using that with a comment explaining why it's used there, or continue further and rename that class to something like `ModInfo`.

Also I'm wondering how it would be possible to not cache star difficulties per setting value without actually dropping the  mod check entirely?

Reply:
> Also I'm wondering how it would be possible to not cache star difficulties per setting value without actually dropping the mod check entirely?

Probably with a change in `MemoryCachingComponent` that can be used to control whether a value should be cached or not. And setting it to not cache any star difficulties for nonstandard settings.

Reply:
Rather than storing the mods directly, could it work by storing hashcodes? Like `mods?.OrderBy(m => m.Acronym).Select(m => m.GetHashCode()).ToArray()`

Reply:
The problem with storing hashcodes is hash collisions, in which a `null` value setting would have the same hash code as a `0` value setting.

Reply:
And how is that different to how it's currently being done?

Reply:
`ConcurrentDictionary` first checks for the hash code of the items to be equal (which can have collisions), then checks for the equality of the items via `Equals` to ascertain they're equal.

The current solution deep-clones the mods so that they return correct hash codes and equality comparsion results.

Your proposed solution would be to store the hashcodes of the mods and use them in both `Lookup.GetHashCode` and `Lookup.Equals`, that would keep the hash collision issue persisting, while it was going to be resolved by checking for `Mod.Equals` on `Lookup.Equals`.

Reply:
Adding more heuristics/fields to the hashcode could practically work but isn't theoretically correct. A hash function is not a unique mapping, it cannot be due to the pigeonhole principle. That's why the dictionary uses equality as fallback in the first place; a collision may never occur, and we'll be none the wiser, or it'll occur again, but only in some even more insanely specific case which will likely be impossible to trace back.

Reply:
Also note that the underlying settings changing value makes the lookup effectively dead even with the hashcode change, because the *equality check* will start failing. So as far as I can see, the cache is actually *not even being used right now* (although I haven't tested this).

Reply:
I thought it shouldn't do that? how do you think the equality check will start failing if it's now deep-cloned from the original mod list?

Reply:
My last comment was referring to the state of things in `master`. Maybe slightly confusing on my behalf not to state that explicitly.

Reply:
Ah, I see.